lib/ansible/plugins/connection/ssh.py

```diff
@@ -396,13 +404,25 @@ def _build_command(self, binary, *other_args):
 
         # If we want to use password authentication, we have to set up a pipe to
         # write the password to sshpass.
-
-        if self._play_context.password:
+        pkcs11_provider = self.get_option("pkcs11_provider")
+        if self._play_context.password or len(pkcs11_provider) >= 1:
             if not self._sshpass_available():
-                raise AnsibleError("to use the 'ssh' connection type with passwords, you must install the sshpass program")
+                raise AnsibleError("to use the 'ssh' connection type with passwords or pkcs11_provider, you must install the sshpass program")
 
+        if self._play_context.password and len(pkcs11_provider) >= 1:
```

Comment:
just add teh 2 extra options in the if, leave the main command construction same (also no 80 col limit please)

Reply:
docs should also reflect this needs a newer version of sshpass, most of the installed ones do not allow for the -p option

Reply:
Maybe i am not understanding, but I have to add the extra options (such as PKCS11Provider) after the ssh binary is added to the command. So it would have to be after line 427 right?

I am trying to track down what is the minimum version of sshpass required to support -P.  Should I do an additional check to verify the version of sshpass supports -P and error if it doesn't (and pkcs11_provider was used).  Do you happen to know the files I need to update the documentation for, this is my first contribution to Ansible so i'm still learning my way around.

Reply:
the docs in the ssh plugin are enough, just add it to the description of the new option

Reply:
I updated the doc to explain the requirement for sshpass 1.6+ and got rid of my 80 col limit, I did format for readability though (example lines 435-438 I formatted so it was easier to read what options were being added).